### PR TITLE
Initial Tracing Implementation

### PR DESCRIPTION
**Pull request by harupy**
```
git pull upstream tracing
```

<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/11454?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11454/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11454
```

</p>
</details>

### Related Issues/PRs

 #xxx

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracing/export/mlflow.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    TRUNCATION_SUFFIX,
+    TraceAttributeKey,
+)
+from mlflow.tracing.types.model import Span, Trace, TraceData, TraceInfo
+from mlflow.tracing.types.wrapper import MLflowSpanWrapper
+
+_logger = logging.getLogger(__name__)
+
+
+class MLflowSpanExporter(SpanExporter):
```
Comment (harupy) [inline mlflow/tracing/export/mlflow.py:21]:
```suggestion
class MlflowSpanExporter(SpanExporter):
```

File: mlflow/tracking/_tracking_service/client.py [hunk 2]
```diff
@@ -144,6 +159,102 @@ def create_run(self, experiment_id, start_time=None, tags=None, run_name=None):
...
+        tags = tags or {}
+
+        return self.store.create_trace(
+            experiment_id=experiment_id,
+            timestamp_ms=timestamp_ms,
+            execution_time_ms=execution_time_ms,
+            status=status,
+            request_metadata=request_metadata or {},
+            tags=tags or {},
+        )
```
Comment (harupy) [inline mlflow/tracking/_tracking_service/client.py:194]: a

File: mlflow/entities/span_event.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    # Use current time if not provided.
+    timestamp: Optional[int] = field(default=int(time.time() * 1e6))
+    attributes: Dict[str, Any] = field(default_factory=dict)
+
+    @classmethod
+    def from_exception(cls, exception: Exception):
+        "Create a span event from an exception."
+
+        stack_trace = cls._get_stacktrace(exception)
+        cls.__init__(
```
Comment (harupy) [inline mlflow/entities/span_event.py:33]: 😳

File: mlflow/entities/trace_data.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+    Args:
+        spans: List of spans that are part of the trace.
+    """
+
+    spans: List[Span] = field(default_factory=list)
+
+    @classmethod
+    def from_dict(cls, d):
+        return cls(spans=[Span(**span) for span in d["spans"]])
```
Comment (harupy) [inline mlflow/entities/trace_data.py:19]:
```suggestion
        return cls(spans=[Span.from_dict(**span) for span in d["spans"]])
```

We might need to use `from_dict` here and cast fields.

File: mlflow/entities/trace.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        See https://ipython.readthedocs.io/en/stable/config/integrating.html#MyObject
+        for more details.
+
+        At the moment, the only supported MIME type is "application/databricks.mlflow.trace",
+        which contains a JSON representation of the Trace object. This object is deserialized
+        in Databricks notebooks to display the Trace object in a nicer UI.
+        """
+        return {
+            "application/databricks.mlflow.trace": self.to_json(),
+            "text/plain": self.__repr__(),
```
Comment (harupy) [inline mlflow/entities/trace.py:40]:
```suggestion
            "text/plain": repr(self),
```

**Comment by serena-ruan** [comment]
@mlflow-automation autoformat

**Pull request merged**